torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -79,31 +87,41 @@ void UnrollPass::handle(Expr* expr) {
 
     non_trivial_pred_found_ = true;
 
+    Expr* expr_with_predicate = expr;
+
     // When a predicate needs to account for ShiftOp, it is currently
     // taken care by its own function.
     if (GpuLower::current()->haloInfo()->needsShiftPredicate(expr)) {
-      ShiftPredicateInserter::insert(
-          expr, for_loops_, thread_pred, unswitched_loop_);
+      expr_with_predicate = ShiftPredicateInserter::insert(
+          expr_with_predicate, for_loops_, thread_pred, unswitched_loop_);
+      if (expr_with_predicate != expr) {
+        registerReplace(expr, expr_with_predicate, &for_loops_.back()->body());
+      }
       return;
     }
 
     // Reduction may need a separate predicate for writes.
-    if (!isReductionInitExpr(expr) && out_tv->domain()->hasReduction()) {
+    if (!isReductionInitExpr(expr_with_predicate) &&
+        out_tv->domain()->hasReduction()) {
       const auto write_pred = unswitched_loop_
           ? thread_pred_expr
           : IrBuilder::create<kir::Predicate>(
-                PredicateType::ReductionWrite, expr, thread_pred);
-      expr->setWritePredicate(write_pred);
+                PredicateType::ReductionWrite,
+                expr_with_predicate,
```

Comment:
It seems a little odd to pass `expr_with_predicate` here as it's not the real expr that we want this predicate to be associated with. This is also related to the comment I made above on the expr linking. Maybe we could just remove the link?

Reply:
Yes, we can use the real `expr`, changed that. See https://github.com/csarofeen/pytorch/pull/2025#discussion_r987564529